drivers/storage/ide/pciidex/fdo.c

```diff
@@ -1,491 +1,507 @@
-		else if (IdentifyData.UltraDMAActive & 0x1)
-			Current = UDMA_MODE0;
-
-		if (IdentifyData.UltraDMASupport & 0x10)
-			Best = UDMA_MODE4;
-		else if (IdentifyData.UltraDMASupport & 0x8)
-			Best = UDMA_MODE3;
-		else if (IdentifyData.UltraDMASupport & 0x4)
-			Best = UDMA_MODE2;
-		else if (IdentifyData.UltraDMASupport & 0x2)
-			Best = UDMA_MODE1;
-		else if (IdentifyData.UltraDMASupport & 0x1)
-			Best = UDMA_MODE0;
-	}
-
-	*BestXferMode = Best;
-	*CurrentXferMode = Current;
-	return TRUE;
+    PAGED_CODE();
+
+    if (FdoExtension->Common.PageFiles ||
+        FdoExtension->Common.HibernateFiles ||
+        FdoExtension->Common.DumpFiles)
+    {
+        Irp->IoStatus.Information |= PNP_DEVICE_NOT_DISABLEABLE;
+    }
+
+    Irp->IoStatus.Status = STATUS_SUCCESS;
+
+    return STATUS_SUCCESS;
 }
 
-static NTSTATUS
-PciIdeXFdoStartDevice(
-	IN PDEVICE_OBJECT DeviceObject,
-	IN PIRP Irp)
+static
+CODE_SEG("PAGE")
+PPDO_DEVICE_EXTENSION
+PciIdeXPdoCreateDevice(
+    _In_ PFDO_DEVICE_EXTENSION FdoExtension,
+    _In_ ULONG ChannelNumber)
 {
-	PPCIIDEX_DRIVER_EXTENSION DriverExtension;
-	PFDO_DEVICE_EXTENSION DeviceExtension;
-	PCM_RESOURCE_LIST ResourceList;
-	NTSTATUS Status;
-
-	DPRINT("PciIdeXStartDevice(%p %p)\n", DeviceObject, Irp);
-
-	DriverExtension = IoGetDriverObjectExtension(DeviceObject->DriverObject, DeviceObject->DriverObject);
-	ASSERT(DriverExtension);
-	DeviceExtension = (PFDO_DEVICE_EXTENSION)DeviceObject->DeviceExtension;
-	ASSERT(DeviceExtension);
-	ASSERT(DeviceExtension->Common.IsFDO);
-
-	DeviceExtension->Properties.Size = sizeof(IDE_CONTROLLER_PROPERTIES);
-	DeviceExtension->Properties.ExtensionSize = DriverExtension->MiniControllerExtensionSize;
-	Status = DriverExtension->HwGetControllerProperties(
-		DeviceExtension->MiniControllerExtension,
-		&DeviceExtension->Properties);
-	if (!NT_SUCCESS(Status))
-		return Status;
-
-	DriverExtension->HwUdmaModesSupported = DeviceExtension->Properties.PciIdeUdmaModesSupported;
-	if (!DriverExtension->HwUdmaModesSupported)
-		/* This method is optional, so provide our own one */
-		DriverExtension->HwUdmaModesSupported = PciIdeXUdmaModesSupported;
-
-	/* Get bus master port base, if any */
-	ResourceList = IoGetCurrentIrpStackLocation(Irp)->Parameters.StartDevice.AllocatedResources;
-	if (ResourceList
-		&& ResourceList->Count == 1
-		&& ResourceList->List[0].PartialResourceList.Count == 1
-		&& ResourceList->List[0].PartialResourceList.Version == 1
-		&& ResourceList->List[0].PartialResourceList.Revision == 1
-		&& ResourceList->List[0].PartialResourceList.PartialDescriptors[0].Type == CmResourceTypePort
-		&& ResourceList->List[0].PartialResourceList.PartialDescriptors[0].u.Port.Length == 16)
-	{
-		DeviceExtension->BusMasterPortBase = ResourceList->List[0].PartialResourceList.PartialDescriptors[0].u.Port.Start;
-	}
-	return STATUS_SUCCESS;
+    NTSTATUS Status;
+    UNICODE_STRING DeviceName;
+    WCHAR DeviceNameBuffer[sizeof("\\Device\\Ide\\PciIde999Channel9-999")];
```

Comment:
According to Printf below `\\Device\\Ide\\PciIde%uChannel%u-%x`, last number is hexadecimal.